File: R/metricsgraphics.R
#' Create a new metricsgraphics.js  plot
#'
#' \code{mjs_plot()} initializes the metricsgraphics.js html widget
#' and takes a data frame & (bare or quoted) x & y column names as minimum input.
#' This must be piped to a "geom" (metricsgraphics.js only supports single
#' "geom" layers) and can also be piped to other \code{mjs_} functions that
#' manipulate aesthetics.
#'
#' See \href{http://metricsgraphicsjs.org/}{MetricsGraphics.js} for more information.
#'
#' @param data data frame
#' @param x bare or quoted name of column to use for x values
#' @param y bare or quoted name of column to use for y values
#' @param show_rollover_text determines whether or not to show any text when a data point is rolled over.
#' @param linked inks together all other graphs whose linked option is set to true.
#'        When one graphic in that set is rolled over, the corresponding values in the other
#'        graphics are also rolled over (default: \code{FALSE} - not linked)
#' @param decimals the number of decimals to show in a rollover (default: \code{2})
#' @param format sets the format of the data object, which is to say,
#'        \code{count}s or \code{percentage}s
#' @param missing_is_hidden if true and if the data object is a time series, missing data points will be treated as zeros
#' @param left the size of the left margin in pixels.
#' @param right the size of the right margin in pixels.
#' @param top the size of the top margin in pixels.
#' @param bottom the size of the bottom margin in pixels.
#' @param buffer the buffer size in pixels between the actual chart area and the margins.
#' @param width Width in pixels (optional, defaults to automatic sizing)
#' @param height Height in pixels (optional, defaults to automatic sizing)
#' @param title plot title
#' @param description plot description
#' @note Plot \code{title} and \code{description} work best when the widget is in
#'       a Bootstrap template. They also increase the overall plot area (height,
#'       mostly) since they add \code{<div>}s. The \code{description} will be
#'       visible in the upper left area (on \code{?} hover) if not displayed
#'       in a Boostrap template.
#' @return metricsgraphics object
#' @export
#' @examples
#' data.frame(year=seq(1790, 1970, 10),
#'            uspop=as.numeric(uspop)) %>%
#'   mjs_plot(x=year, y=uspop) %>%
#'   mjs_line()
#'
#' # accessor params can also be quoted
#'
#' data.frame(year=seq(1790, 1970, 10),
#'            uspop=as.numeric(uspop)) %>%
#'   mjs_plot(x="year", y="uspop") %>%
#'   mjs_line()
#'
mjs_plot <- function(data, x, y,
                     show_rollover_text = TRUE,
                     linked = FALSE,
                     decimals=2, format="count",
                     missing_is_hidden=FALSE,
                     left = 80, right = 10,
                     top = 40, bottom = 60, buffer = 8,
                     width = NULL, height = NULL,
                     title=NULL, description=NULL) {

  if (!format %in% c("percentage", "count")) {
    stop("'format' must be either 'percentage' or 'count'")
  }

  if (inherits(data, "data.frame")) data <- data.frame(data, stringsAsFactors=FALSE)

  eid <- sprintf("mjs-%s",
                 paste(sample(c(letters[1:6], 0:9), 30, replace=TRUE), collapse=""))

  if (!missing(x)) {
    x <- substitute(x)
    res <- try(eval(x, data, parent.frame()), silent = TRUE)
    if (!inherits(res, "try-error") && inherits(res, "character")) {
      if (length(res) != 1) {
        x <- as.character(substitute(x))
      } else {
        x <- res
      }
    } else if (inherits(x, "name")) { x <- as.character(x) }
  } else {
    x <- as.character(substitute(x))
  }

  if (!missing(y)) {
    y <- substitute(y)
    res <- try(eval(y, data, parent.frame()), silent = TRUE)
    if (!inherits(res, "try-error") && inherits(res, "character")) {
      if (length(res) != 1) {
        y <- as.character(substitute(y))
      } else {
        y <- res
      }
    } else if (inherits(y, "name")) { y <- as.character(y) }
  } else {
    y <- as.character(substitute(y))
  }

  is_datetime <- function(x) {
    inherits(x, c('Date', 'POSIXct', 'POSIXlt'))
  }

  is_posix <- function(x) {
    inherits(x, c('POSIXct', 'POSIXlt'))
  }

  orig_posix <- FALSE
  if (is.null(dim(data))) {
    if (is_posix(data)) orig_posix <- TRUE
  } else if (is_posix(data[, x])) {
    orig_posix <- TRUE
  }

  if (is.null(dim(data))) {
    if (is_datetime(data)) data <- as.numeric(data)
  } else if (is_datetime(data[, x])) {
    data[, x] <- as.numeric(data[, x])
  }

  params = list(
    forCSS=NULL,
    regions=NULL,
    orig_posix=orig_posix,
    data=data,
    x_axis=TRUE,
    y_axis=TRUE,
    baseline_accessor=NULL,
    predictor_accessor=NULL,
    show_confidence_band=NULL,
    show_secondary_x_label=NULL,
    chart_type="line",
    xax_format="plain",
    x_label=NULL,
    y_label=NULL,
    markers=NULL,
    baselines=NULL,
    linked=linked,
    title=title,
    description=description,
    left=left,
    right=right,
    bottom=bottom,
    buffer=buffer,
    format=format,
    y_scale_type="linear",
    yax_count=5,
    xax_count=6,
    x_rug=FALSE,
    y_rug=FALSE,
    area=FALSE,
    missing_is_hidden=missing_is_hidden,
    size_accessor=NULL,
    color_accessor=NULL,
    color_type="number",
    color_range=c("blue", "red"),
    size_range=c(1, 5),
    bar_height=20,
    min_x=NULL,
    max_x=NULL,
    min_y=NULL,
    max_y=NULL,
    bar_margin=1,
    binned=FALSE,
    bins=NULL,
    least_squares=FALSE,
    interpolate="cardinal",
    decimals=decimals,
    show_rollover_text=show_rollover_text,
    x_accessor=x,
    y_accessor=y,
    multi_line=NULL,
    geom="line",
    yax_units="",
    legend=NULL,
    legend_target=NULL,
    y_extended_ticks=FALSE,
    x_extended_ticks=FALSE,
    target=sprintf("#%s", eid)
  )

  if (is.null(width)) params$full_width <- TRUE
  if (is.null(height)) params$full_height <- TRUE

  htmlwidgets::createWidget(
    name = 'metricsgraphics',
    x = params,
    width = width,
    height = height,
    package = 'metricsgraphics',
    elementId = eid
  )

}

#' Widget output function for use in Shiny
#'
#' @param outputId output id
#' @param width width
#' @param height height
#' @export
metricsgraphicsOutput <- function(outputId, width = '100%', height = '400px'){
  htmlwidgets::shinyWidgetOutput(outputId, 'metricsgraphics', width, height, package = 'metricsgraphics')
}

#' Widget render function for use in Shiny
#'
#' @param expr expr
#' @param env env
#' @param quoted quoted
#' @export
renderMetricsgraphics <- function(expr, env = parent.frame(), quoted = FALSE) {
  if (!quoted) { expr <- substitute(expr) } # force quoted
  htmlwidgets::shinyRenderWidget(expr, metricsgraphicsOutput, env, quoted = TRUE)
}

metricsgraphics_html <- function(id, style, class, ...) {
  list(tags$div(id = id, class = class, style = style),
       tags$div(id = sprintf("%s-legend", id), class = sprintf("%s-legend", class)))
}

File: inst/htmlwidgets/metricsgraphics.js
HTMLWidgets.widget({

  name: 'metricsgraphics',

  type: 'output',

  initialize: function(el, width, height) {

    return {  }

  },

  renderValue: function(el, params, instance) {

    // save params for reference from resize method
    instance.params = params;

    // draw the graphic
    this.drawGraphic(el, params, el.offsetWidth, el.offsetHeight);

  },

  drawGraphic: function(el, params, width, height) {

    // remove existing children
    while (el.firstChild)
      el.removeChild(el.firstChild);

    dbg = params

    wide = null;

    is_multi_line = params.y_accessor instanceof Array;

    if (params.geom == "histogram") {
      wide = params.data;
    } else {

      wide = HTMLWidgets.dataframeToD3(params.data);

      if (is_multi_line) {

        tmp = [];

        for (var i=0; i<params.y_accessor.length; i++) {
          var data = {};
          data["value"]= params.data[params.y_accessor[i]];
          data[params.x_accessor]  = params.data[params.x_accessor];
          tmp.push(HTMLWidgets.dataframeToD3(data));
        }

        wide = tmp ;

      }
    }

    var xax_format = mjs_plain;

    if (params.xax_format == "date") {

      xax_format = mjs_date ;

      if (!is_multi_line) {
        MG.convert.date(wide, params.x_accessor)
      } else {
        for (var i=0; i<wide.length; i++) {
          wide[i] = MG.convert.date(wide[i], params.x_accessor);
        }
      }

      if (params.markers != null) {
        for (var i=0; i<params.markers.length; i++) {
          params.markers[i][params.x_accessor] =
            d3.time.format("%Y-%m-%d").parse(params.markers[i][params.x_accessor]);
        }
      }

    }

    if (params.xax_format == "comma") xax_format = mjs_comma ;

    var mg_params = {

        data: wide,
        target: '#' + el.id,
        xax_format: xax_format,

        x_axis: params.x_axis,
        y_axis: params.y_axis,

        interpolate: params.interpolate,
        decimals: params.decimals,
        format: params.format,

        color_type: params.color_type,
        color_range: params.color_range,
        point_size: params.point_size,
        size_range: params.size_range,

        markers: params.markers,
        baselines: params.baselines,

        show_rollover_text: params.show_rollover_text,

        y_scale_type: params.y_scale_type,

        chart_type: params.chart_type,
        x_accessor: params.x_accessor,
        y_accessor: params.y_accessor,
        color_accessor: params.color_accessor,
        size_accessor: params.size_accessor,

        show_rollover_text: params.show_rollover_text,

        legend: params.legend,
        legend_target: params.legend_target,

        bar_margin: params.bar_margin,
        binned: params.binned,
        bins: params.bins,

        linked: params.linked,

        height: height,
        width: width,
        bottom: params.bottom,
        top: params.top,
        right: params.right,
        left: params.left,
        buffer: params.buffer,

        area: params.area,
        animate_on_load: params.animate_on_load,
        y_rug: params.y_rug,
        x_rug: params.x_rug,

        min_x: params.min_x,
        max_x: params.max_x,
        min_y: params.min_y,
        max_y: params.max_y,

        yax_count: params.yax_count,
        xax_count: params.xax_count,

        least_squares: params.least_squares,

        x_label: params.x_label,
        y_label: params.y_label,
        title: params.title,
        description: params.description

    };

    if (!is_multi_line) {

      MG.data_graphic(mg_params);

    } else {

      mg_params.y_accessor = "value";

      if (!params.grid) {

        //draw a multiline chart
        MG.data_graphic(mg_params);

      } else {

        var n = params.y_accessor.length;

        // arrange sizes
        mg_params.width  = width/n;
        mg_params.height = height/n;
        mg_params.bottom = params.bottom/n + 10;
        mg_params.top    = params.top/n + 10;
        mg_params.right  = params.right/n;
        mg_params.left   = params.left/n;
        mg_params.buffer = params.buffer/n;
        mg_params.yax_count = Math.ceil(params.yax_count/n);
        mg_params.xax_count = Math.ceil(params.xax_count/n);

        mg_params.linked = true;

        var parentDom = document.querySelector(mg_params.target);

        for (var i=0; i<n; i++) {
          var y_accessor = params.y_accessor[i];
          var div = document.createElement("div");
          div.id = y_accessor;
          div.style.float = "left";
          parentDom.appendChild(div);

          mg_params.data   = wide[i];
          mg_params.target = "#" + y_accessor;

          // add titles instead of legends
          if (mg_params.legend instanceof Array){
            mg_params.title = params.legend[i];
          }
          mg_params.legend_target = null;

          //draw a separate chart
          MG.data_graphic(mg_params);
        }

      }
    }

  },

  resize: function(el, width, height, instance) {
    if (instance.params)
      this.drawGraphic(el, instance.params, width, height);
  }

});

mjs_comma = function(d) {
  var df =  d3.format("0,000");
  return df(d);
}

mjs_date = function(d) {
  var df = d3.time.format('%b %d');
  return df(d);
}

mjs_plain = function(d) {
 return(d.toString());
}
